docs/src/writing_good_rules.md

```diff
@@ -35,6 +63,77 @@ Examples being:
 - There is only one derivative being returned, so from the fact that the user called
   `frule`/`rrule` they clearly will want to use that one.
 
+## Structs: constructors and functors
+
+To define an `frule` or `rrule` for a _function_ `foo` we dispatch on the type of `foo`, which is `typeof(foo)`.
+For example, the `rrule` signature would be like:
+
+```julia
+function rrule(::typeof(foo), args...; kwargs...)
+    ...
+    return y, foo_pullback
+end
+```
+
+For a struct `Bar`,
+```julia
+struct Bar
+    a::Float64
+end
+
+(bar::Bar)(x, y) = return bar.a + x + y # functor
+```
+we can define an `frule`/`rrule` for the `Bar` constructor(s), as well as any `Bar` [functors](https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects).
+
+To define an `rrule` for a constructor for a  _type_ `Bar` we need to be careful to dispatch only on `Type{Bar}`.
+For example, the `rrule` signature for a `Bar` constructor would be like:
+```julia
+function ChainRulesCore.rrule(::Type{Bar}, a)
```

Comment:
I think Seth's point about `<:` (for non-concrete types) is worth making too
https://github.com/JuliaDiff/ChainRulesCore.jl/pull/331#discussion_r606984609

Reply:
in general probably should always use `<:` inside `Type`. 
Since it means your code doesn't have to change if it gains type-parameters.
(possibly even this should be added to BlueStyle)